src/Core/Content/Seo/SeoUrlPlaceholderHandler.php

```diff
@@ -49,10 +66,18 @@ public function replace(string $content, string $host, SalesChannelContext $cont
 
             if (preg_match_all('/' . self::DOMAIN_PLACEHOLDER . '[^#]*#/', $content, $matches)) {
                 $mapping = $this->createDefaultMapping($matches[0]);
-                $seoMapping = $this->createSeoMapping($context, $mapping);
-                foreach ($seoMapping as $key => $value) {
-                    $seoMapping[$key] = $host . '/' . ltrim($value, '/');
-                }
+                $key = $this->generateKey($context, $mapping);
+
+                $seoMapping = $this->cache->get($key, function (ItemInterface $item) use ($context, $mapping, $host, $key) {
```

Comment:
This change does make the application more complicated and hurts more than it helps. The given mapping is highly individual and won't be used by any other page (as they will include other links) and you will have the pain of invalidating this cache.

Additionally, this kind of caching happens when the page is in HTTP cache.

Reply:
@shyim I completly changed the PR now.

I updated the old query builder to now use subqueries instead. This method is much faster and takes not even 1/5 the time of the old query.